Adding info for grouping step definitions

zutshiy opened the pull request:
Added a new section to help newcomers understand the power of abstracting steps and using helper methods/classes to allow for more scalable and maintainable suites in cucumber.

zutshiy committed:
Adding info for abstracting steps with helpers

mlvandijk commented on content/cucumber/step-definitions.md:
````diff
@@ -253,3 +253,48 @@ Given("I have {int} {color} balls", function (int1, color) {
 ```
 {{% /block %}}
+
+# Abstract Your Steps
````
I'd prefer to not use the term "abstract" (to avoid confusion with the java keyword).
Maybe it should be called "helper methods" - which is fact might be a better location: https://docs.cucumber.io/gherkin/step-organization/#helper-methods
And as it is asked more often, add a question "How do I group my step definitions?" with a brief answer and link to this info.

zutshiy replied:
Didn't understand you completely. Do you want me to move this to the **helper-methods** section and link to there from the **step-definitions** page? Or the other way round?

mlvandijk replied:
Merge with the helper-method section please. :)

mlvandijk replied:
Oh, turns out we already have a section on Helper methods in the FAQ too: https://docs.cucumber.io/community/faq/#how-to-use-helper-methods

So what I meant before was, there should be a question in the FAQ page (there is), which should give a brief answer to the question (it does) and point to the relevant section in the docs (https://docs.cucumber.io/gherkin/step-organization/#helper-methods). Like we do for all the other FAQ.

zutshiy replied:
Alright. I have made some updates here.
Once we finalize the content, I ll move it to the new section and link it from the FAQs.

mlvandijk replied:
Content is fine; I suggest merging with the other section; we might need some edits if there's any duplication. :)

zutshiy replied:
Ofcourse. Let me merge this with that section then.

zutshiy replied:
Ok so i updated the FAQ and actually removed the example from there to keep it brief. We already provide a good example in the `group step definitions` section, so I removed it from the FAQ.

If you think we should keep it, I would suggest putting it inside the `Helper methods` section of the Step Organization tab instead. To keep all the info regarding helper methods at one place.

mlvandijk commented on content/cucumber/step-definitions.md:
````diff
@@ -253,3 +253,48 @@ Given("I have {int} {color} balls", function (int1, color) {
+
+with the step def :
+
+```java
````
We'll need to add examples for the other languages too. I can do Kotlin, we ask help for javascript & ruby if needed.

zutshiy replied:
Yes that would be great :)

I am only familiar with Java so I didn't write one for the other three.

mlvandijk replied:
That's fine. We'll get there.
You could add code blocks for javascript, ruby, kotlin with a `// TODO` inside and we'll get to them later.

zutshiy replied:
Done

mlvandijk commented on content/cucumber/step-definitions.md:
```diff
@@ -253,3 +253,48 @@ Given("I have {int} {color} balls", function (int1, color) {
+
+You can use the same method to write steps for validating a webpage, clicking a button, etc each with their own helper methods.
+
+A good look at the _(Abstract) Factory Design_ pattern can be very useful for creating/refractoring your test cases with such abstraction. Also, using [Data Tables](/cucumber/api/#data-tables) for providing inputs along with this method makes your steps even more easy to maintain and understand.
```
A link to a good explanation of this pattern might be helpful.

zutshiy replied:
Will do. I was actually thinking about it but wasn't sure if thats allowed.

mlvandijk replied:
We can link to external sites, but prefer to do so "sparingly". If you'd like some more guidelines check the [contributing to documentation](https://docs.cucumber.io/community/contributing-to-documentation/) page. Please let me know if anything is unclear!

zutshiy replied:
Thank You :) I ll have a look. For now I have added a link

zutshiy committed:
Fixing Lists

mlvandijk commented on content/cucumber/step-definitions.md:
````diff
@@ -253,3 +253,48 @@ Given("I have {int} {color} balls", function (int1, color) {
+    Given I get the contact details
+```
+
+and all of these steps have different step definitions which visit the Home, About and Contact pages, you might be writing _redundant steps_. It is possible that the underlying code for executing each of these test cases is different, **but** it is important to understand that from the point of view of their **behavior**, all of these steps do essentially similar things, ie.
````
Please use British English "behaviour" for consistency.

zutshiy replied:
Sure

mlvandijk commented on content/cucumber/step-definitions.md:
```diff
@@ -253,3 +253,48 @@ Given("I have {int} {color} balls", function (int1, color) {
+This helps you in a number of ways like,
+
+1. Allowing you to have less and more easily maintainable steps.
+2. Making your project easily scalable: Adding tests for a new functionality with the same underlying _behavior_ is a lot easier and less cumbersome.
```
Same here ("behaviour")

zutshiy commented on content/cucumber/step-definitions.md:
```diff
@@ -253,3 +253,48 @@ Given("I have {int} {color} balls", function (int1, color) {
+
+_* Open the Home page_
+_* Open the About page_
+_* Open the Contact page_
```
Ok, I can't figure out how to make this a list.

mlvandijk replied:
Try removing the underscores, and if needed add a white line between items.

zutshiy replied:
Great. I ll try that.

zutshiy replied:
Finally.

mlvandijk replied:
The joys of markdown ;)

mlvandijk commented on content/cucumber/step-definitions.md:
```diff
@@ -253,3 +253,48 @@ Given("I have {int} {color} balls", function (int1, color) {
+
+# Abstract Your Steps
+
+It is always a good idea to make sure that you are not writing duplicate/similar step definitions. While there are a number of ways to do so and you can document your step definitions as well, it is always good to make use of **helper methods** to abstract your step definitions and reduce the number of steps.
```
The content is good, but could be more concise. See if you can use less words to convey the same information. For example (but not limited to):
"It is always a good idea to make sure that you are not writing duplicate/similar step definitions." -> "Make sure that you are not writing very similar step definitions."

zutshiy replied:
Made it more concise. I ll keep checking and see if i can update it more, but take a look meanwhile, please :)

mlvandijk commented on content/cucumber/step-definitions.md:
````diff
@@ -253,3 +253,48 @@ Given("I have {int} {color} balls", function (int1, color) {
+}
+```
+
+And then inside your step definition, use a helper method to determine how the page must be opened and direct your step definition to the correct code for opening that particular page.
````
Not part of our Slack discussion, but triggered by your example:
1. There is a balance to consider between reuse of step definitions and helper method, and (over)complicating the implementation. If you get to the point where you need unit tests for your test implementation, it's too complicated ;) In fact, some people prefer step definitions without variables to avoid (more) complicated implementation of step definitions.

2. Personally I use the "helper methods" to "talk to my application", be it the UI / page objects, or the API. This keeps my step definitions "thin" (only providing the glue between Gherkin steps and my code.

zutshiy replied:
I did want to portray using the step definitions only as a "glue" between Gherkin and the code. Maybe I wasn't able to convey that properly here. If you could help here, that would be great.

Ah, maybe saying "use a helper method" caused that issue, I should replace that with "Use a factory builder to call the needed code" or something similar for better understanding.

Let me know what you think. You are free to update it as well :)

mlvandijk replied:
Let me have a look.

zutshiy replied:
Ok, so i removed this section and added a more simpler sentence to convey the message properly. Let me know if its better :)

mlvandijk reviewed (commented):
Thanks for getting started on this @gazler22! I do have some suggestions, if you don't mind. I'm happy to collaborate if you like :)

zutshiy committed:
Update step-definitions.md

mlvandijk commented on content/cucumber/step-definitions.md:
```diff
@@ -253,3 +253,60 @@ Given("I have {int} {color} balls", function (int1, color) {
+
+Avoid writing similar step definitions which can lead to clutter. While documenting your steps helps, making use of **helper methods** to abstract them can do wonders.
+
+For eg, if you have the following steps :
```
Please change "eg" -> "example" (for consistency)

mlvandijk replied:
Much more concise! Very nice :)

zutshiy replied:
Sure. And thank you :)

mlvandijk commented on content/cucumber/step-definitions.md:
````diff
@@ -253,3 +253,60 @@ Given("I have {int} {color} balls", function (int1, color) {
+    Given I get the contact details
+```
+
+and all of these steps visit the **Home**, **About** and **Contact** pages, you might be writing _redundant steps_. While the underlying code for these steps could be different, their **behaviour** is essentially the same, ie.
````
You can use one `*` if you want text to be *italic*, and two `**` if you want text to be **bold**. While underscores may do same / similar, this will help with consistency.

mlvandijk commented on content/cucumber/step-definitions.md:
````diff
@@ -253,3 +253,60 @@ Given("I have {int} {color} balls", function (int1, color) {
+ //TODO
+```
+
+And have your step definitions act as a gateway to the actual code (in this case, a factory method to decide which page to open).
````
"And have your step definitions act as a gateway to the actual code (in this case, a factory method to decide which page to open)." -> Suggestion: "Your step definitions are the glue to the actual code (in this example, a factory method to decide which page to open)."

You could also have a look at the example in the [FAQ](https://docs.cucumber.io/community/faq/#how-to-use-helper-methods); it goes nicely with yours. You might also want to have step definitions that hide a bunch of implementation details (like how to log in) and/or combine a few of these reusable methods (login, go to page, etc).

zutshiy replied:
I was going to add a line to emphasize "hiding implementation details by having one step call several helper methods" as well but the FAQ already covers it pretty well so I decided against it.

mlvandijk replied:
It would be my preference to clearly explain in the docs themselves, and provide a summary of key points in the FAQ, pointing to the relevant section(s) of the docs from there. So a little duplication wouldn't hurt imho. But let's see what we end up with. (Incrementally "refactor" the docs :))

zutshiy replied:
Alright. I ll put it in then. Let me know what you think.

mlvandijk commented on content/cucumber/step-definitions.md:
```diff
@@ -253,3 +253,60 @@ Given("I have {int} {color} balls", function (int1, color) {
+
+You can use the same method to write steps for validating a webpage, clicking a button, etc each with their own helper methods.
+
+We recommend taking a look at the [Factory Design Pattern] (https://refactoring.guru/design-patterns/factory-method) as well.
```
I'd prefer "suggest" over "recommend", as I have no strong feelings on this.

zutshiy replied:
Yes that makes sense. Updated.

mlvandijk commented:
There should be an option where I can propose edits for you to accept, but I cannot find it.

zutshiy commented:
> 
> 
> There should be an option where I can propose edits for you to accept, but I cannot find it.

That's fine. Just let me know here and I ll update it.

mlvandijk committed:
Edits for consistency / clarity

zutshiy committed:
Update step-definitions.md

zutshiy commented:
Wait, sorry. I think I overrode your changes. Let me update the file with both of our changes.

zutshiy committed:
Update step-definitions.md

zutshiy committed:
Update step-definitions.md

mlvandijk commented:
> Wait, sorry. I think I overrode your changes. Let me update the file with both of our changes.

Yeah, that's why I was looking for the suggest edit feature. I know it exists (others have used it in this project), I just can't find it.

zutshiy commented on content/cucumber/step-definitions.md:
```diff
@@ -253,3 +253,55 @@ Given("I have {int} {color} balls", function (int1, color) {
+
+1. Less and more easily maintainable steps.
+2. Increased Scalablility: Adding tests for a new functionality with the same underlying *behaviour* is easier.
+3. Concise and easy to understand tests.
```
Do you think it'd be better to put these in a single sentence instead?
The list is starting to look a little annoying to me now 😄

mlvandijk replied:
Use bullit points instead? And make them more concise. Eg:
```
* Better maintainability of steps
* Easier reuse of steps
* Easier to understand
```

zutshiy replied:
Sure

zutshiy committed:
Removing section from step-definitions

zutshiy committed:
Adding Group step defs section

Moved from step-definitions section to step-organization. Updated list for advantages.

zutshiy committed:
Updateing FAQ to link to Grouping step defs

mpkorstanje commented:
@mlvandijk 

![image](https://user-images.githubusercontent.com/6946919/55282205-da2df200-533f-11e9-8665-60a243ee6a32.png)

The left most button when adding a line comment.

mlvandijk commented on content/community/faq.md:
```diff
@@ -146,24 +146,10 @@ Each scenario should *test exactly one thing* so that when it fails, it fails fo
-We will illustrate the use of helper methods with an example. For instance, let's say you need to login to your application.
-If you want to (re)use this action as a step or part of another step, write a method for login.
+Helper methods allow you to re(use) actions in steps and avoid writing multiple step definitions for similar behaviours. You write a helper method for actions like *logging in to your application* and then reuse them as a step or part of other steps.
+This helps keep your tests clean, concise and maintainable.
```
+1 very good

mlvandijk committed:
Minor edits

mlvandijk commented:
Thnx @mpkorstanje - I never would have found that

mlvandijk commented:
Thanks for this @gazler22 !

zutshiy commented on content/gherkin/step-organization.md:
```diff
@@ -76,23 +76,23 @@ This makes your project a lot easier to understand for people who join your proj
     Given I go to the {} page
 
-with the step definition :
+And the following the step definition:
```
there's an extra "the" here, i ll raise a PR to fix the typo.
